google-compute-engine/src/main/java/org/jclouds/googlecomputeengine/compute/GoogleComputeEngineServiceAdapter.java

```diff
@@ -70,223 +74,234 @@
-         credentials.password(options.getLoginPassword());
-      }
-      if (options.shouldAuthenticateSudo() != null) {
-         credentials.authenticateSudo(options.shouldAuthenticateSudo());
-      }
-      return credentials.build();
-   }
-
-   private static String toName(URI link) {
-      String path = link.getPath();
-      return path.substring(path.lastIndexOf('/') + 1);
-   }
+public final class GoogleComputeEngineServiceAdapter implements ComputeServiceAdapter<Instance, MachineType, Image, Location> {
+
+	private final JustProvider							justProvider;
+	private final GoogleComputeEngineApi				api;
+	private final Resources								resources;
+	private final Map<URI, URI>							diskToSourceImage;
+	private final Predicate<AtomicReference<Operation>>	operationDone;
+	private final Predicate<AtomicReference<Instance>>	instanceVisible;
+	private final FirewallTagNamingConvention.Factory	firewallTagNamingConvention;
+	private final List<String>							imageProjects;
+
+
+	@Inject GoogleComputeEngineServiceAdapter(JustProvider justProvider, GoogleComputeEngineApi api, Predicate<AtomicReference<Operation>> operationDone,
+			Predicate<AtomicReference<Instance>> instanceVisible, Map<URI, URI> diskToSourceImage, Resources resources,
+			FirewallTagNamingConvention.Factory firewallTagNamingConvention, @Named(IMAGE_PROJECTS) String imageProjects) {
+		this.justProvider = justProvider;
+		this.api = api;
+		this.operationDone = operationDone;
+		this.instanceVisible = instanceVisible;
+		this.diskToSourceImage = diskToSourceImage;
+		this.resources = resources;
+		this.firewallTagNamingConvention = firewallTagNamingConvention;
+		this.imageProjects = Splitter.on(',').omitEmptyStrings().splitToList(imageProjects);
+	}
+
+
+	@Override public NodeAndInitialCredentials<Instance> createNodeWithGroupEncodedIntoName(String group, String name, Template template) {
+		GoogleComputeEngineTemplateOptions options = GoogleComputeEngineTemplateOptions.class.cast(template.getOptions());
+		checkNotNull(options.network(), "template options must specify a network");
+		checkNotNull(template.getHardware().getUri(), "hardware must have a URI");
+		checkNotNull(template.getImage().getUri(), "image URI is null");
+
+		List<AttachDisk> disks = Lists.newArrayList();
+		disks.add(AttachDisk.newBootDisk(template.getImage().getUri()));
+
+		disks.addAll(options.getDisks());
+		if (null != options.getAutoCreateDiskOptions()) {
+			AutoCreateDiskOptions diskOptions = options.getAutoCreateDiskOptions();
+			checkArgument(template.getLocation().getScope() == LocationScope.ZONE);
+			DiskApi diskApi = api.disksInZone(template.getLocation().getId());
+			Operation op = diskApi.create(diskOptions.getDiskName(name), new DiskCreationOptions.Builder().sizeGb(diskOptions.diskSizeGb).build());
+			// TODO: Is there a way to avoid the op, and just AttachDisk.create? where do we get the URI from? does it have to pre-exist?
```

Comment:
I don't think there is a way to avoid the operation. You have to create the disk before attaching it.

https://cloud.google.com/compute/docs/instances#startinstanceapi